### DISCUSSION TITLE
How to delete the same substructure in one molecule separately

### DISCUSSION THREAD

**Discussion by JuChunHuang** [rdkit/rdkit#4685] [Q&A]
Hello everyone! I would like to delete substructure in a molecule in order to modify this molecule. I have a question. Suggesting that there are two same substructures that I want to delete in a molecule. How can I delete each substructure at one time?

This is my code:

>   if mol.HasSubstructMatch(patt):
    flag = 1
    mol = Chem.DeleteSubstructs(mol, patt)
    mol_frags = Chem.GetMolFrags(mol, asMols=True)
    frag_size=[]
    for x in mol_frags:
      frag_size.append(x.GetNumAtoms())
    largest_frag_index = frag_size.index(max(frag_size))
    largest_frag = mol_frags[largest_frag_index]
  else:
    flag = 0
    largest_frag = mol

For example, if I want to delete the Cl group in `CSc1nc2ccccn2c(=N)c1S(=O)(=O)c1ccc(Cl)c(Cl)c1`
![image](https://user-images.githubusercontent.com/81243182/141237678-01a70652-13bf-4b0e-980a-b99240457d5a.png)
It will lead to the deletion of all Cl groups. Is it possible to obtain a molecule that is without the first Cl group, and a molecule that is without the second Cl group? Thank you very much.

**Comment by greglandrum** [comment]
I can come up with two straightforward ways to do this.

Directly removing the atoms:
```
Cl_pattern = Chem.MolFromSmarts('[ClD1]')
def atom_remover(mol,pattern=Cl_pattern):
    matches = mol.GetSubstructMatches(pattern)
    if not matches:
        yield Chem.Mol(mol)
    for match in matches:
        res = Chem.RWMol(mol)
        res.BeginBatchEdit()
        for aid in match:
            res.RemoveAtom(aid)
        res.CommitBatchEdit()
        Chem.SanitizeMol(res)
        yield res
```
You call this like `ms = [x for x in atom_remover(mol)]`

Or you can use a chemical reaction:
```
cl_rxn = AllChem.ReactionFromSmarts('[ClD1:1]-[*:2]>>[H:1]-[*:2]')
def atom_remover2(mol,rxn=cl_rxn):
    ps = rxn.RunReactants((mol,))
    if not ps:
        yield Chem.Mol(mol)
    for p in ps:
        res = Chem.RemoveHs(p[0])        
        yield res
 ```
You call this like `ms = [x for x in atom_remover2(mol)]`

The second approach is more robust, because it will work with aromatic heterocycles. To see this, try calling either with the input SMILES `'Cln1cccc1'`

There's a notebook here with the same code:
https://gist.github.com/greglandrum/12bd952b61c98225c9b82840e652174a

**Reply by JuChunHuang** [reply]
Sorry for the late reply. Thank you very much for this information!
Since I am a newbie, I wonder if I can ask some simple questions.

1. If I need to remove functional groups(e.g., carboxyl), would it be better to use the chemical reaction method?
2. Regarding the chemical reaction approach, am I supposeed to know every atomic number in the molecule so that I can remove the specific functional group (if there are two same functional groups in the molecule)?

Thank you!

**Reply by greglandrum** [reply]
The chemical reaction approach is much more flexible than other methods, so I think it's a good one to use.

I don't understand your second question; can you please give a specific example?

**Reply by JuChunHuang** [reply]
> The chemical reaction approach is much more flexible than other methods, so I think it's a good one to use.
> 
> I don't understand your second question; can you please give a specific example?

I wonder regarding the chemical reaction approach, if I want to remove a functional group (not an atom), it seems to me that I have to be aware of every atom id so that I can use `AllChem.ReactionFromSmarts('here')`.

**Reply by greglandrum** [reply]
I'm still not quite sure what you're asking. Here, for example, is a reaction which replaces carboxylic acid groups with an H:
```
rxn = AllChem.ReactionFromSmarts('[*:1]-C(=O)[OH]>>[*:1]-[H]')
```
That doesn't require knowing atom IDs in advance, but can be applied to remove carboxylic acid groups one at a time from any molecule.

**Reply by adelenelai** [reply]
Hi @JuChunHuang, I proposed a possible solution for removing molecular substructures (so, not atoms) in my latest [blogpost](https://adelenel.ai/deleteonesubstructure/). 

Could in theory work for FGs if you define them correctly in the query.

**Comment by adelenelai** [comment]
Hi again, I had forgotten about this Discussion and in the mean time came up with a slightly different way of removing substructure matches and returning the fragments.

For example if you want to remove all possible 'CC' matches from this molecule (highlighted in pink):

![image](https://github.com/rdkit/rdkit/assets/25338941/ee9e1f53-bea2-4510-801e-86359d3d80b3)

You call
`remove_substr_matches_return_frag("CCNC(=O)[C@@H]1C=C2c3cccc4[nH]cc(c34)C[C@H]2N(C)C1","CC")`
and get these remaining fragments

![image](https://github.com/rdkit/rdkit/assets/25338941/64d0991c-8fa3-4e79-b89a-ce2c1a9d2cdc)


You can also express the substructure in SMARTS.
See the code here:
https://adelenel.ai/remove-substr-iter/
